src/app/public/theming/theme.service.ts

```diff
@@ -0,0 +1,134 @@
+import {
+  Injectable,
+  Renderer2
+} from '@angular/core';
+
+import {
+  Observable,
+  ReplaySubject
+} from 'rxjs';
+
+import {
+  SkyThemeSettings
+} from './theme-settings';
+
+import {
+  SkyThemeSettingsChange
+} from './theme-settings-change';
+
+/**
+ * Provides methods for updating and handling changes to the current theme.
+ */
+@Injectable()
+export class SkyThemeService {
+
+  /**
+   * Notifies consumers when the current theme has changed.
+   */
+  public get themeChange(): Observable<SkyThemeSettingsChange> {
+    return this.themeObs;
+  }
+
+  private current: SkyThemeSettings;
+
+  private hostEl: any;
+
+  private renderer: Renderer2;
+
+  private theme: ReplaySubject<SkyThemeSettingsChange>;
+
+  private themeObs: Observable<SkyThemeSettingsChange>;
+
+  constructor() { }
+
+  /**
+   * Initializes the theme service with the specified parameters. This should only be called
+   * from a host component that provides its own theme to child components.
+   * @param hostEl The host element under which themed components are rendered.
+   * @param renderer A Renderer2 instance for updating the host element with theme changes.
+   * @param theme The initial theme.
+   */
+  public init(
+    hostEl: any,
+    renderer: Renderer2,
+    theme: SkyThemeSettings
+  ): void {
+    this.hostEl = hostEl;
+    this.renderer = renderer;
+
+    this.theme = new ReplaySubject<SkyThemeSettingsChange>(1);
+    this.themeObs = this.theme.asObservable();
+
+    this.setTheme(theme);
+  }
+
+  /**
+   * Destroys the current theme service. This should only be called from a host component that
+   * provides its own theme to child components.
+   */
+  public destroy(): void {
+    this.theme.complete();
+
+    this.hostEl =
+      this.renderer =
+      undefined;
+  }
+
+  /**
+   * Updates the current theme.
+   * @param newTheme The new theme to apply.
+   */
+  public setTheme(newTheme: SkyThemeSettings): void {
```

Comment:
Would it be possible to combine both `init` and `setTheme` methods into one, to avoid confusion and to have only one point of interaction with the service?

Another option would be to return an "instance" from the `init()` method that has a public `setTheme` method on it, similar to our modal and overlay services. Just a thought.

"Themer" is a dumb name, but to illustrate:
```
const themer = this.themeService.createThemer(this.elementRef, this.renderer);
themer.setTheme(...);
this.themeService.destroy(themer);
```

(The intention for the change would be to make it difficult for the consumer to run `setTheme` before `init`.)

Reply:
The problem with both approaches is I want a child component to be able to update the theme without a reference to the host element. The service needs to be initialized by the host component one time so that child components can set their parent component's theme. This will make more sense when we add controls for enabling/disabling dark mode or toggling themes in documentation where an individual element and its children are themed rather than the whole page..